### PR TITLE
Improved testability of DirectoryFinder

### PR DESCRIPTION
**Pull request by x789**
Instead of using `System.IO`, the `DirectoryFinder` uses a small abstraction layer to access the file-system. This allows its logic to be tested thoroughly in unit tests that are not dependent on the real file system.

More about the _how_ and _why_ can be found in the comments of issue #789.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/nunit/nunit-console?pullRequest=855) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/nunit/nunit-console?pullRequest=855) before we can accept your contribution.<br/><br/>:x: x789 [sign now](https://cla.dotnetfoundation.org/nunit/nunit-console?pullRequest=855)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/nunit/nunit-console?pullRequest=855) it.</sub>

**Comment by x789** [comment]
As was to be expected -- build breaks on Linux/macOS...
Will work on it... 🧐👨‍💻

**Comment by x789** [comment]
Fixed tests so they pass on non-Windows systems.
-> PR is ready for review...

File: src/NUnitEngine/nunit.engine.tests/Internal/FileSystemAccess/Default/DirectoryTests.cs
Comment (mikkelbu) [inline src/NUnitEngine/nunit.engine.tests/Internal/FileSystemAccess/Default/DirectoryTests.cs:71]: Couldn't this be simplified to
```suggestion
            var path = SIO.Directory.GetCurrentDirectory() + SIO.Path.DirectorySeparatorChar;
```

File: src/NUnitEngine/nunit.engine.tests/Internal/DirectoryFinderTests.cs
Comment (mikkelbu) [inline src/NUnitEngine/nunit.engine.tests/Internal/DirectoryFinderTests.cs:187]:
```suggestion
            var expected = new[]
            {
                CombinePath(baseDir.FullName),
                CombinePath(baseDir.FullName, "addins"),
                CombinePath(baseDir.FullName, "tests"),
                CombinePath(baseDir.FullName, "tests", "v1"),
                CombinePath(baseDir.FullName, "tests", "v1", "tmp"),
                CombinePath(baseDir.FullName, "tests", "v2"),
                CombinePath(baseDir.FullName, "addins", "morph"),
                CombinePath(baseDir.FullName, "addins", "empty")
            };
```

File: src/NUnitEngine/nunit.engine.tests/Internal/DirectoryFinderTests.cs
Comment (mikkelbu) [inline src/NUnitEngine/nunit.engine.tests/Internal/DirectoryFinderTests.cs:298]: Perhaps this would be easier to read if we extracted `new IDirectory[] { this.GetFakeDirectory("tools", "metamorphosator", "tests", "v1"), this.GetFakeDirectory("tools", "metamorphosator", "tests", "v2") }` into a variable?

File: src/NUnitEngine/nunit.engine.tests/Internal/DirectoryFinderTests.cs
Comment (mikkelbu) [inline src/NUnitEngine/nunit.engine.tests/Internal/DirectoryFinderTests.cs:387]:
```suggestion
            var expected = new[]
            {
                this.GetFakeDirectory("tools", "frobuscator", "tests"),
                this.GetFakeDirectory("tools", "frobuscator", "tests", "abc"),
                this.GetFakeDirectory("tools", "frobuscator", "tests", "def"),
                this.GetFakeDirectory("tools", "metamorphosator", "tests"),
                this.GetFakeDirectory("tools", "metamorphosator", "tests", "v1"),
                this.GetFakeDirectory("tools", "metamorphosator", "tests", "v1", "tmp"),
                this.GetFakeDirectory("tools", "metamorphosator", "tests", "v2")
            };
```

File: src/NUnitEngine/nunit.engine.tests/Internal/DirectoryFinderTests.cs
Comment (mikkelbu) [inline src/NUnitEngine/nunit.engine.tests/Internal/DirectoryFinderTests.cs:539]:
```suggestion
            var expected = new[]
            {
                this.GetFakeFile("tools", "frobuscator", "tests", "abc", "tests.abc.dll"),
                this.GetFakeFile("tools", "frobuscator", "tests", "abc", "tests.123.dll"),
                this.GetFakeFile("tools", "frobuscator", "tests", "def", "tests.def.dll"),
                this.GetFakeFile("tools", "metamorphosator", "tests", "v1", "test-assembly.dll"),
                this.GetFakeFile("tools", "metamorphosator", "tests", "v1", "test-assembly.pdb"),
                this.GetFakeFile("tools", "metamorphosator", "tests", "v2", "test-assembly.dll"),
                this.GetFakeFile("tools", "metamorphosator", "tests", "v2", "test-assembly.pdb")
            };
```

**Review by mikkelbu — APPROVED** [review]
Great work @x789 👍. I've left some comments/suggestions, but overall this looks like a great addition. Thanks for the work.

File: src/NUnitEngine/nunit.engine.tests/Internal/FileSystemAccess/Default/FileTests.cs
Comment (x789) [reply src/NUnitEngine/nunit.engine.tests/Internal/FileSystemAccess/Default/FileTests.cs]: I'm afraid there is no guarantee.
I changed the code so this test will be ignored if `invalidCharThatIsNotInInvalidPathChars` contains the default because the test will not make sense.

File: src/NUnitEngine/nunit.engine.tests/Internal/FileSystemAccess/Default/FileSystemTests.cs
Comment (x789) [reply src/NUnitEngine/nunit.engine.tests/Internal/FileSystemAccess/Default/FileSystemTests.cs:161]: The tests need the path of the test-assembly. Therefore, `TestContext.CurrentContext.TestDirectory` does not work.

File: src/NUnitEngine/nunit.engine.tests/Internal/FileSystemAccess/Default/DirectoryTests2.cs
Comment (x789) [reply src/NUnitEngine/nunit.engine.tests/Internal/FileSystemAccess/Default/DirectoryTests2.cs]: Good point. Removed it. 😀

File: src/NUnitEngine/nunit.engine.tests/Internal/DirectoryFinderTests.cs
Comment (x789) [reply src/NUnitEngine/nunit.engine.tests/Internal/DirectoryFinderTests.cs]: Yes. I extracted the content of `testsDir` as suggested to a variable. Did the same for `baseDir` too...

File: src/NUnitEngine/nunit.engine.core/IEnumerableExtensions.cs
Comment (x789) [reply src/NUnitEngine/nunit.engine.core/IEnumerableExtensions.cs]: Sorry, I didn't pay attention to that at all. I have now moved the usings in all files so that they fit the project style.

File: src/NUnitEngine/nunit.engine.core/Internal/DirectoryFinder.cs
Comment (x789) [reply src/NUnitEngine/nunit.engine.core/Internal/DirectoryFinder.cs]: Me neither. But 'where' sounds good to me, too.

File: src/NUnitEngine/nunit.engine.core/Internal/FileSystemAccess/Default/Directory.cs
Comment (x789) [reply src/NUnitEngine/nunit.engine.core/Internal/FileSystemAccess/Default/Directory.cs]: I was unsure about it because the development guide and the files in the project differ.
I fixed the preamble in all new files.

**Comment by x789** [comment]
Thank you @mikkelbu for the review and for finding the flaws in the code. I just updated the PR according to your suggestions.

File: src/NUnitEngine/nunit.engine.core/Internal/FileSystemAccess/Default/Directory.cs
Comment (CharliePoole) [reply src/NUnitEngine/nunit.engine.core/Internal/FileSystemAccess/Default/Directory.cs]: @x789 It would be useful for us to know where you found conflicting info, so we can fix it.

Comment (x789) [reply src/NUnitEngine/nunit.engine.core/Internal/FileSystemAccess/Default/Directory.cs]: @CharliePoole 
The [NUnit Developer Info](https://docs.nunit.org/articles/developer-info/NUnit-Copyright-Notice.htm) says:

> // Copyright (c) yyyy Charlie Poole

The file [ProcessUtils.cs](https://github.com/nunit/nunit-console/blob/master/src/NUnitEngine/nunit.engine.core/Internal/ProcessUtils.cs) contains:

> // Copyright (c) 2016 Joseph N. Musser II

My chain of thought was then as follows:
- It is important to have the license and a copyright claim in the header. 
- The developer-info-page stated "Charlie Poole" as an example.
- The reason why "Charlie Poole, Rob Prouse" is written in all other files is that they initially created them.

Comment (CharliePoole) [reply src/NUnitEngine/nunit.engine.core/Internal/FileSystemAccess/Default/Directory.cs]: @x789 I can see how that was confusing all right.

Our practice (apparently not documented anywhere) is to put the copyright holders for the __whole application__ in each file. (There has been some discussion about changing this, but that's a separate matter.)

Currently, the copyright holders for the nunit-console project are myself and @rprouse .

The example file you cited, with copyright listed as @jnm2 is an anomaly we should deal with. The documentation page that merely lists a copyright block without any text explaining what it means needs to be fixed or deleted as well. I'll create some separate issues.

Sorry for our confusing you but thanks for having found these anomalies!

File: src/NUnitEngine/nunit.engine.tests/Internal/DirectoryFinderTests.cs
Comment (mikkelbu) [inline src/NUnitEngine/nunit.engine.tests/Internal/DirectoryFinderTests.cs:32]: Last nit-pick :). These should be moved outside. Besides this the rest looks great. Again thanks for the contribution. We need a second review before merging this.

**Comment by ChrisMaddock** [comment]
Thanks for doing this @x789, it looks great! I'm just working on getting the next release out this weekend, then next weekend I'll come back to this and give it the second review, and hopefully get it merged in. 😄

File: src/NUnitEngine/nunit.engine.core/Internal/DirectoryFinder.cs
Comment (ChrisMaddock) [inline src/NUnitEngine/nunit.engine.core/Internal/DirectoryFinder.cs:108]: Thanks for taking this out. 🙂

File: src/NUnitEngine/nunit.engine.tests/Internal/FileSystemAccess/Default/DirectoryTests.cs
Comment (ChrisMaddock) [inline src/NUnitEngine/nunit.engine.tests/Internal/FileSystemAccess/Default/DirectoryTests.cs:203]: How about using NUnit's PlatformAttribute instead of this method? It should just be possible to add `[Platform("Win")]` to the necessary tests - just a bit tidier, perhaps. 🙂

**Review by ChrisMaddock — COMMENTED** [review]
Hi @x789 - thanks for doing all this - it's a great improvement on what we had! 

Sorry it's taken a while for me to take a look through, but everything here looks good. Just left a couple of minor suggestions - see what you think, and we'll get this merged in. 😄

File: src/NUnitEngine/nunit.engine.core/IEnumerableExtensions.cs
Comment (x789) [reply src/NUnitEngine/nunit.engine.core/IEnumerableExtensions.cs]: Moved 🙂

File: src/NUnitEngine/nunit.engine.tests/Internal/FileSystemAccess/Default/DirectoryTests.cs
Comment (x789) [reply src/NUnitEngine/nunit.engine.tests/Internal/FileSystemAccess/Default/DirectoryTests.cs:203]: I initially wanted to use the `Platform` attribute. But it is not available when the target framework is `netcoreapp1.1`.
I don't like my approach with the private method either, but I think it's better than a combination of a method to support .NET Core 1.1 and an attribute (for all other platforms).

If we want to make the exclusion more readable by using an attribute, I see two possibilities:
1. Add support for .NET Core 1.1 to `PlatformAttribute` (this would require a change in NUnit.Framework and we will need to use that version).
2.We create a new internal attribute (e.g. `WindowsOnlyAttribute`) that works for all target frameworks and contains the rudimental logic that is currently contained in `DirectoryTests.SkipOnNonWindowsSystems()`.

Comment (ChrisMaddock) [reply src/NUnitEngine/nunit.engine.tests/Internal/FileSystemAccess/Default/DirectoryTests.cs:203]: Ah sorry - we've just removed the netcoreapp1.1 target recently, it must have been since you started this PR. Can you merge in master, and then you should just be able to add it in?

Comment (x789) [reply src/NUnitEngine/nunit.engine.tests/Internal/FileSystemAccess/Default/DirectoryTests.cs]: Ooops, I missed the removal of netcoreapp1.1.
Replaced my OS detection with the `PlatformAttribute`. Works like a charm! 😀

**Comment by ChrisMaddock** [comment]
Great - thanks for making those last changes. 

I'll merge this now - thanks for doing it all, it's in a much better state than it was before! 🙂

**Pull request merged**
